### PR TITLE
refactor(router): Add a state manager to manage router URL state

### PR DESCRIPTION
**Pull request by atscott**
This commit refactors the router internals to track state inside a separate
`StateManager`. This helps open the door to managing the state in
different, swappable providers. The current interface needed by the
Router is: `currentUrlTree`, `rawUrlTree`, `browserUrlTree`,
`routerState`, and `handleNavigationEvent` (other properties are because
some router properties are writeable when they shouldn't be). This is
a suprisingly small interface and can hopefully be made smaller in the
future (i.e. removing 1 or more of the 3 tracked of `UrlTree`s).

Diffbase: #48427

### CODE REVIEW

**Comment by atscott** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/559275283/OCL:559275283:BASE:561057615:1693327496604:337e649;groups=PossiblyNewlyFailing/targets) (1 failure due to access to private variable)

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

File: packages/router/src/navigation_transition.ts [hunk 1]
```diff
@@ -266,7 +272,7 @@ interface InternalRouterInterface {
   // All of these are public API of router interface and can change during runtime because they are
   // writeable. Ideally, these would be removed and the values retrieved instead from the values
   // available in DI.
-  errorHandler: ErrorHandler;
+  errorHandler: (error: any) => any;
```
Comment (AndrewKushnir) [inline packages/router/src/navigation_transition.ts:275]: Just curious why we changed types here?

File: packages/router/src/state_manager.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    if (this.canceledNavigationResolution !== 'computed') {
+      return this.currentPageId;
+    }
+    return (this.location.getState() as RestoredState | null)?.ɵrouterPageId ?? this.currentPageId;
+  }
+  routerState: RouterState;
+  private stateMemento = this.createStateMemento();
+
+  constructor() {
+    this.currentUrlTree = new UrlTree();
```
Comment (AndrewKushnir) [inline packages/router/src/state_manager.ts:114]: nit: can we move those fields from the constructor to particular field initializers instead?

File: goldens/size-tracking/integration-payloads.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
   "cli-hello-world-lazy": {
     "uncompressed": {
       "runtime": 2734,
-      "main": 232018,
+      "main": 237908,
```
Comment (AndrewKushnir) [inline goldens/size-tracking/integration-payloads.json:19]: This is likely an accumulated delta over time, I don't see reasons why this change would cause 5K+ size increase.

**Review by AndrewKushnir — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

File: goldens/size-tracking/integration-payloads.json (hunk 3 above)
Comment (atscott) [reply goldens/size-tracking/integration-payloads.json:19]: Yea, I would guess accumulated delta, especially since there aren't any new symbols being picked up in this change. But I also thought it was the old 500B limit so this is a lot more alarming.

File: packages/router/src/navigation_transition.ts (hunk 1 above)
Comment (atscott) [reply packages/router/src/navigation_transition.ts:275]: I'm trying to get rid of ErrorHandler completely. The name overlaps with the one from core so it caused problems here: https://github.com/angular/angular/issues/47212. So it was already removed from ExtraOptions it just wasn't removed from the Router API.

File: packages/router/src/state_manager.ts (hunk 2 above)
Comment (atscott) [reply packages/router/src/state_manager.ts:114]: Done.

File: goldens/size-tracking/integration-payloads.json (hunk 3 above)
Comment (atscott) [reply goldens/size-tracking/integration-payloads.json:19]: Okay, I verified that there was no unexpected new code pulled in due to this change.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking

**Comment by atscott** [comment]
caretaker note: Probably should merge and sync on its own

**Comment by jessicajaniuk** [comment]
@atscott Looks like the size-tracking changed since other stuff merged in and now there's a conflict.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 4555290d54629b3fb8b17aa600492802f98df680.
